consensus/src/data_manager.rs

```diff
@@ -0,0 +1,231 @@
+                pos,
+                pos.digest(),
+                logical_time
+            );
+            if logical_time <= pos.expiration() {
+                receivers.push(
+                    self.data_reader
+                        .load()
+                        .as_ref()
+                        .unwrap() //TODO: can this be None? Need to make sure we call new_epoch() first.
+                        .get_batch(pos)
+                        .await,
+                );
+            } else {
+                debug!("QS: skipped expired pos");
+            }
+        }
+        receivers
+    }
+
+    pub async fn notify_commit(&self, logical_time: LogicalTime, payloads: Vec<Payload>) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            self.data_reader
+                .load()
+                .as_ref()
+                .unwrap()
+                .update_certified_round(logical_time)
+                .await;
+
+            let digests: Vec<HashValue> = payloads
+                .into_iter()
+                .flat_map(|payload| match payload {
+                    Payload::DirectMempool(_) => {
+                        unreachable!()
+                    }
+                    Payload::InQuorumStore(proof_with_status) => proof_with_status.proofs,
+                    Payload::Empty => Vec::new(),
+                })
+                .map(|proof| *proof.digest())
+                .collect();
+
+            let _ = self
+                .quorum_store_wrapper_tx
+                .load()
+                .as_ref()
+                .unwrap()
+                .as_ref()
+                .clone()
+                .try_send(WrapperCommand::CleanRequest(logical_time, digests));
+        }
+    }
+
+    pub async fn update_payload(&self, block: &Block) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) && block.payload().is_some() {
+            match block.payload().unwrap() {
+                Payload::InQuorumStore(proof_with_status) => {
+                    if proof_with_status.status.lock().is_none() {
+                        let receivers = self
+                            .request_data(
+                                proof_with_status.proofs.clone(),
+                                LogicalTime::new(block.epoch(), block.round()),
+                            )
+                            .await;
+                        proof_with_status
+                            .status
+                            .lock()
+                            .replace(DataStatus::Requested(receivers));
+                    }
+                }
+                Payload::Empty => {}
+                Payload::DirectMempool(_) => {
+                    unreachable!()
+                }
+            }
+        }
+    }
+
+    //Assumses it is never called for the same block concurrently. Otherwise status can be None.
+    pub async fn get_data(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
```

Comment:
I'm not too fond of get_payload since we already have a payload in the block. We are actually taking a payload and extracting transactions from it.
Then how about just get_transactions(()?